RFC: Execution Environments (#274)
buildpacks/rfcs: text/0000-execution-environments.md

## Setting the Execution Enviroment for Build + Buildpack
A platform will set the `CNB_EXEC_ENV` env var to the execution environment desired. Buildpacks can than read this env var to branch or switch on logic needed based on the execution environment.

In addition, Builder Authors, Buildpack Authors, and App Developers will be able to specify various options to a specific execution enviroment using the `exec-env` key.

## Development Environments
The specifics of creating development enviroments are out of scope of this RFC, but it's not hard to extrapolate how these kind of changes can assist in creating Buildpacks for development environments.

# How it Works
[how-it-works]: #how-it-works

## `exec-env` key in TOML

In order to support additional execution environments an `exec-env` key will be added to various TOML tables in the project. The value can be any string with `all` having special meaning. `all` will apply to all execution environments and will be the default if not specified. This should make it backwards compatible and optional. When `exec-env` is not set to `all`, the table settings will only be applied to that execution environment.

> ### Project Descriptor - `project.toml` (App Developers)

Reviewer: Can you add an example of a full project.toml that is used for producing a test image and a production image?

Author: Just for Project Descriptor?

# Meta
[meta]: #meta
- Name: Execution Environments
- Start Date: 2023-01-11
- Author(s): @hone
- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)

# Summary
[summary]: #summary

Add support for different execution environments for buildpacks, with testing as the first use case.

# Definitions
[definitions]: #definitions

- Execution Environment - The target environment the OCI image is expected to be run in, i.e. production, test, development.

# Motivation
[motivation]: #motivation

> The main way Buildpacks are being used is building production images, but this is only one piece of the CI/CD pipeline. Without a solid buildpack test environment story, users will be required alternatives for building their environment. A testing environment, while different, shares many of the same broad strokes ultimately producing an execution environment. This is something Buildpacks are well suited to solve.

Reviewer 2: ```suggestion
The main way Buildpacks are being used is building production images, but this is only one piece of the software development process. Without a solid buildpack test environment story, users will be required alternatives for building their environment. A testing environment, while different, shares many of the same broad strokes ultimately producing an execution environment. This is something Buildpacks are well suited to solve.
```

I'm thinking this RFC could be useful outside of CI/CD pipelines, most notably for testing changes locally. However, I could also imagine building development environments locally with features like hot reload, auto recompilation, web servers that respond with stack traces, etc.

# What it is
[what-it-is]: #what-it-is

## Test Environments

One of the goals in this design is to minimize the changes needed. The Buildpack lifecycle is already well suited to produce environments:

```
App Source -> OCI Image (Production Environment) -> Execute (Launch Process)
```

To create a test environment, it can follow the same flow:

```
App Source -> OCI Image (Test Environment) -> Execute (Test Process)
```

### Division of Responsibility
With the test OCI Image, a platform can execute the tests in the pipeline as they see fit. This means a bulk of the responsibilities are platform concerns:

- Set which environment to build for

> - Decide which buildpacks to execute

Reviewer 3: Isn't this a lifecycle concern? i.e., when processing a group within an order, should it skip buildpacks that do not declare `exec-env` matching the desired env?

Author: When talking to @jabrown85, I thought the Platform/Builder provide the `order.toml` to `lifecycle`?

- How to execute the tests
- What is the test result format like [TAP](https://en.wikipedia.org/wiki/Test_Anything_Protocol)?
- How to process the test results

> - What to do with the results

Reviewer 3: Is this ever something we would want to spec? Having a consistent way for buildpacks to e.g., dump test output could help ensure portability across platforms.

Author: I could see us wanting to do that, but I wasn't sure how much we wanted to impose standards there.

This narrows the scope of the Buildpack responsibilities to what it's already good at:

- Produce Test Environment
- How to launch the process
  - launch.toml with test process marked as default (recommendation)

## Setting the Execution Enviroment for Build + Buildpack
A platform will set the `CNB_EXEC_ENV` env var to the execution environment desired. Buildpacks can than read this env var to branch or switch on logic needed based on the execution environment.

In addition, Builder Authors, Buildpack Authors, and App Developers will be able to specify various options to a specific execution enviroment using the `exec-env` key.

## Development Environments
The specifics of creating development enviroments are out of scope of this RFC, but it's not hard to extrapolate how these kind of changes can assist in creating Buildpacks for development environments.

> # How it Works

Reviewer 3: It would be nice to see the pack flow as well

Reviewer 3: Should we add anything to the app image to designate it as being built for a particular environment? To avoid users accidentally deploying a test image in production...

I could see folks wanting to use the same tag when re-building a test image for production, in order to use previously cached dependencies.

Author: @natalieparellano that's a good question. Do you think cached dependencies should be shared b/t different execution environments?

Author: Is a label the best place to designate the execution environment?

Author: For sharing cache, is it a problem if I run a test, get dev dependencies but my production build doesn't want them and prunes that cache. Next time I run a test I would need to install them. Do you just want to share cache for bootstrapping for every build?

Reviewer 4: I don't feel like that will be an issue for things like the JVM or Node.js runtime, but it sounds like it would be for other layers, like node modules or other things installed via package manager that would change between environments. That would certainly be annoying and not expected as a user.

Caching is per layer, what if the buildpack were to create different layers for cases like this? Maybe incorporate the exec env key into the layer name or something to make it unique? It wouldn't need to do this everywhere, but for things that are env dependent, it could have `npm_modules-test` and `npm_modules-production` layers.

If you ran a build for test and then a build for production, would that result in keeping both of those cached? or would it evict the test layer when you run the production build because the production image didn't use that layer?

[how-it-works]: #how-it-works

## `exec-env` key in TOML

> In order to support additional execution environments an `exec-env` key will be added to various TOML tables in the project. The value can be any string with `all` having special meaning. `all` will apply to all execution environments and will be the default if not specified. This should make it backwards compatible and optional. When `exec-env` is not set to `all`, the table settings will only be applied to that execution environment.

Reviewer 2: What would consume this new key? If a platform or user ran `pack build -e CNB_EXEC_ENV=test my-app:test`, a buildpack might contribute new launch entries flagged for test, but how would a platform or user determine which of the launch entries were `all` or `test`? My guess is at the very least, we'd want this key to show up in the `io.buildpacks.build.metadata` `process-types` data?

Author: yep, good call out.

> In order to support additional execution environments an `exec-env` key will be added to various TOML tables in the project. The value can be any string with `all` having special meaning. `all` will apply to all execution environments and will be the default if not specified. This should make it backwards compatible and optional. When `exec-env` is not set to `all`, the table settings will only be applied to that execution environment.

Reviewer: We should consider wanting multiple values. This might make `[]` mean `all`, which is kind of odd.

```
exec-env: ["staging", "production"] # but not uat
```

Author: I ended up going with the `*` any stack magic value.

### Project Descriptor - `project.toml` (App Developers)

This file be extended by adding `exec-env` to the following tables:

`[[io.buildpacks.group]]`
`[[io.buildpacks.pre.group]]`
`[[io.buildpacks.post.group.env]]`
`[[io.buildpacks.build.env]]`

An example would look like this:

```TOML
[_]
schema-version = "0.3"

[[io.buildpacks.group]]
id = "buildpacks/ruby"
version = "latest"

[[io.buildpacks.group]]
id = "buildpacks/metrics-agent"
version = "latest"
exec-env = "production"

[[io.buildpacks.group]]
id = "buildpacks/headless-chrome"
version = "latest"
exec-env = "test"

[[io.buildpacks.post.group]]
id = "buildpacks/procfile"
version = "latest"

> [[io.buildpacks.buildenv]]

Reviewer 3: ```suggestion
[[io.buildpacks.build.env]]
```

name = "RAILS_ENV"
value = "production"
exec-env = "production"

[[io.buildpacks.build.env]]
name = "RAILS_ENV"
value = "test"
exec-env = "test"

[[io.buildpacks.build.env]]
name = "PARALLEL_WORKERS"
value = "4"
exec-env = "test"
```

### `builder.toml` (Builder Authors)

> The only table that `exec-env` will be added to is `[[order.group.env]]`.

Reviewer 3: Do you mean `[[order.group]]`?

### `buildpack.toml` (Buildpack Authors)

The only table that `exec-env` will be added to is `[[buildpack.order.group]]`. This only is applicable for composite buildpacks.

## `CNB_EXEC_ENV` Environment Variable

This env var will reserve the following values:

* production
* test
* development

> ### Buildpack API

Reviewer: Can we perhaps get a conceptual example of what a buildpack author might do with this new execution environment information?

I have in my head a language buildpack author may...
- install test group dependencies
- create `ruby-tests` process / `ruby-tests-verbose` process
- Skip cleaning up things they might clean up otherwise for production images
- Set env vars to test mode operation (RAILS_ENV or similar)

# Prior Art
[prior-art]: #prior-art

## [Heroku Testpack API](https://devcenter.heroku.com/articles/testpack-api)

As part of the classic buildpack design, Heroku extended the API to include support for tests. It added a `bin/test-compile` phase, which is like the normal `bin/compile` phase but specific for setting things up for test. `bin/test` was called for executing tests. This design let a buildpack have a specific codepath for tests and also do intelligent logic for executing tests based on the source code. For example [in Ruby](https://github.com/heroku/heroku-buildpack-ruby/blob/main/bin/support/ruby_test#L62-L76), the buildpack author can run `bin/rspec` if it detects `rspec`, `bin/rails test` when using rails, or default to `rake test`.

There were some flaws in this design. Though it's clean to separate production and test code paths, they end up sharing a lot of code. Many of the bash based Heroku buildpacks would just [call `bin/compile`](https://github.com/heroku/heroku-buildpack-nodejs/blob/main/bin/test-compile#L24) with different parameters/env vars.

> ## [GOOGLE_DEVMODE](https://cloud.google.com/docs/buildpacks/service-specific-configs#google_devmode)

Reviewer 5: Similar to devmode, I just ran across live reloading in the Paketo python buildpacks https://paketo.io/docs/howto/python/#using-bp_live_reload_enabled

These are specific to the Google Cloud Buildpacks for setting a development code path to work with skaffold.

## [Develop API](https://github.com/buildpacks/spec/pull/71)

The original Cloud Native Buildpacks spec included a Develop API, but it was never implemented.

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

> - "env" is overloaded as a word since we also use it for environment variables. Is there a better word here?

Reviewer 3: How about "mode"?

Reviewer 2: I like "context", but that's just as overloaded. How about "purpose = test" or "intent = test"?

Reviewer 6: I think "env" was overloaded before we got here, and it's ok to use.

- Should there be builders that are specific to an execution environment? What about `include` or `exclude`?
- Should the execution environments be an enum or flexible as a string?
  - enums will help encourage standardization across buildpacks and platforms.
  - strings can help account for use cases we haven't thought of yet.

> - Should buildpacks be allowed specify allowlist execution environments?

Reviewer 3: Are there any downsides to doing this? It would be more flexible and possibly avoid some duplication within orders

Author: I think the only downside is what if a user wants to override in a way it wasn't intended? Is there a reason to lock stuff out?

```
App Source -> OCI Image (Production Environment) -> Execute (Launch Process)
```

To create a test environment, it can follow the same flow:

```
App Source -> OCI Image (Test Environment) -> Execute (Test Process)
```

### Division of Responsibility
With the test OCI Image, a platform can execute the tests in the pipeline as they see fit. This means a bulk of the responsibilities are platform concerns:

- Set which environment to build for
- Decide which buildpacks to execute
- How to execute the tests
- What is the test result format like [TAP](https://en.wikipedia.org/wiki/Test_Anything_Protocol)?
- How to process the test results
- What to do with the results

This narrows the scope of the Buildpack responsibilities to what it's already good at:

- Produce Test Environment
- How to launch the process
  - launch.toml with test process marked as default (recommendation)

## Setting the Execution Enviroment for Build + Buildpack
A platform will set the `CNB_EXEC_ENV` env var to the execution environment desired. Buildpacks can than read this env var to branch or switch on logic needed based on the execution environment.

> In addition, Builder Authors, Buildpack Authors, and App Developers will be able to specify various options to a specific execution enviroment using the `exec-env` key.

Reviewer 6: ```suggestion
In addition, Builder Authors, Buildpack Authors, and App Developers will be able to specify various options to a specific execution enviroment using the `exec-env` key in a `project.toml` file.
```

Author: There are now a few other files like `builder.toml` and `launch.toml`.

```

### `builder.toml` (Builder Authors)

The the only table `exec-env` will be added to is `[[order.group]]` and `[[build.env]]`.

### `buildpack.toml` (Buildpack Authors)

The only table that `exec-env` will be added to is `[[buildpack.order.group]]`. This only is applicable for composite buildpacks.

### `launch.toml` (Buildpack Authors)

Not all process types make sense for every execution environment. In order to help hint to the platform the intention how a process should be used, the `exec-env` key will be added to the `[[processes]]` table.

### `metadata.toml` (Platform Operators)

On the platform side, the `exec-env` key will be added to `metadata.toml` in the `[[processes]]` table to mirror `launch.toml`. This will help platforms to make decisions on which processes are desired for each execution environment. `lifecycle` will list all processes and will not exclude any based on the execution environment Instead, a platform can use the available information to make a decision.

## `CNB_EXEC_ENV` Environment Variable

The spec will reserve the following values to help standardize execution environments:

* production
* test
* development

### Buildpack API

> A buildpack author will be able to determine the execution environment their buildpack is expected to build for by reading the `CNB_EXEC_ENV` environment variable. If this value is not set, a Buildpack Author can assume it's set to `production`. This will be provided for both `bin/detect` and `bin/build`.

Reviewer 6: Instead of defaulting to `production`, should we have a concept of `all`? Maybe this is similar to the `exec-env` having `*`?

Author: @jkutner do we need a concept of "all" for an individual lifecycle execution?

Author: Talking with @jabrown85 , I think compiled languages like Go might want to remove the source tree from `/workspace` for production but not test which would extra complicate things.

## [GOOGLE_DEVMODE](https://cloud.google.com/docs/buildpacks/service-specific-configs#google_devmode)

These are specific to the Google Cloud Buildpacks for setting a development code path to work with skaffold.

## [Develop API](https://github.com/buildpacks/spec/pull/71)

The original Cloud Native Buildpacks spec included a Develop API, but it was never implemented.

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

- "env" is overloaded as a word since we also use it for environment variables. Is there a better word here?
- Should there be builders that are specific to an execution environment? What about `include` or `exclude`?
- Should the execution environments be an enum or flexible as a string?
  - enums will help encourage standardization across buildpacks and platforms.
  - strings can help account for use cases we haven't thought of yet.
- Should buildpacks be allowed specify allowlist execution environments?

> - What changes are needed in the buildpack registry?

Reviewer 4: Users will probably want a way to tell if a buildpack supports execution environments and if so, which ones. Maybe searching by that too. i.e. filter by buildpacks that support execution environments.

Author: @jkutner should we make changes to the registry index to account for this somehow? Add it into `buildpack.toml`?

# Meta
[meta]: #meta
- Name: Execution Environments
- Start Date: 2023-01-11
- Author(s): @hone
- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)

# Summary
[summary]: #summary

> Add support for different execution environments for buildpacks, with testing as the first use case.

Reviewer 7: ```suggestion
Add buildpack support for multiple execution environments, with better support for testing as the first use case.
```

# Definitions
[definitions]: #definitions

> - Execution Environment - The target environment the OCI image is expected to be run in, i.e. production, test, development.

Reviewer 7: ```suggestion
- Execution Environment - The target environment the OCI image is expected to be executed in, i.e. production, test, development.
```

# Motivation
[motivation]: #motivation

> The main way Buildpacks are being used is building production images, but this is only one piece of the software development process. Without a solid buildpack test environment story, users will be required use alternatives for building their environment. A testing environment, while different, shares many of the same broad strokes ultimately producing an execution environment. This is something Buildpacks are well positioned to solve.

Reviewer 7: ```suggestion
Buildpacks are mainly being used to build images for production environments, but this is only one piece of the software development process. Without test environment support for buildpacks, users currently have to implement alternatives for testing their source code. Buildpacks are well positioned to help bridge that gap, as building production and test environments often share many of the same broad strokes.
```

# What it is
[what-it-is]: #what-it-is

## Test Environments

One of the goals in this design is to minimize the changes needed. The Buildpack lifecycle is already well suited to produce environments:

```
App Source -> OCI Image (Production Environment) -> Execute (Launch Process)
```

To create a test environment, it can follow the same flow:

```
App Source -> OCI Image (Test Environment) -> Execute (Test Process)
```

### Division of Responsibility
With the test OCI Image, a platform can execute the tests in the pipeline as they see fit. This means a bulk of the responsibilities are platform concerns:

- Set which environment to build for
- Decide which buildpacks to execute
- How to execute the tests
- What is the test result format like [TAP](https://en.wikipedia.org/wiki/Test_Anything_Protocol)?
- How to process the test results
- What to do with the results

This narrows the scope of the Buildpack responsibilities to what it's already good at:

- Produce Test Environment
- How to launch the process
  - launch.toml with test process marked as default (recommendation)

## Setting the Execution Enviroment for Build + Buildpack

> A platform will set the `CNB_EXEC_ENV` env var to the execution environment desired. Buildpacks can than read this env var to branch or switch on logic needed based on the execution environment.

Reviewer 7: ```suggestion
A platform will set the `CNB_EXEC_ENV` env var to the desired execution environment. Buildpacks can then read this env var to branch or switch on logic needed for the execution environment.
```

> In addition, Builder Authors, Buildpack Authors, and App Developers will be able to specify various options to a specific execution enviroment using the `exec-env` key.

Reviewer 7: ```suggestion
In addition, Builder Authors, Buildpack Authors, and App Developers will be able to configure various options for specific execution environments using the `exec-env` key.
```

## Development Environments

> The specifics of creating development enviroments are out of scope of this RFC, but it's not hard to extrapolate how these kind of changes can assist in creating Buildpacks for development environments.

Reviewer 7: ```suggestion
The specifics of creating development environments are out of scope of this RFC, but it's not hard to extrapolate how the proposed changes can assist in creating Buildpacks for development environments.
```

# How it Works
[how-it-works]: #how-it-works

## `exec-env` key in TOML

> In order to support additional execution environments an `exec-env` key will be added to various TOML tables in the project. This will be an array that takes string values. An individual element can be any string with `*` having special meaning. Similar to the ["any stack RFC"](https://github.com/buildpacks/rfcs/blob/main/text/0056-any-stack-buildpacks.md) `*` will apply to all execution environments. `["*"]` will be the default if not specified. This should make the key backwards compatible and optional. When `exec-env` is not empty and does not include `*`, the table settings will only be applied to the specificed execution environments.

Reviewer 7: ```suggestion
In order to support additional execution environments, an optional `exec-env` key will be added to various TOML tables in the project. This will be an array that takes string values. An individual element can be any string with `*` having special meaning. Similar to the ["any stack RFC"](https://github.com/buildpacks/rfcs/blob/main/text/0056-any-stack-buildpacks.md), `*` will apply to all execution environments. `["*"]` will be the default if not specified. This should make the key backwards compatible. When `exec-env` is not empty and does not include `*`, the table settings will only be applied to the specified execution environments.
```

### Project Descriptor - `project.toml` (App Developers)

> An app developer may have execution environment configuration like only using a metrics agent in production or headless user agent needs in test. In order to facilitate these needs, the project descriptor will be extended by adding `exec-env` to the following tables:

Reviewer 7: ```suggestion
An app developer may have execution environment specific configuration, like only using a metrics agent in production, or a headless user agent in test. In order to facilitate these needs, the project descriptor will be extended by adding `exec-env` to the following tables:
```

* `[[io.buildpacks.group]]`
* `[[io.buildpacks.pre.group]]`
* `[[io.buildpacks.post.group.env]]`
* `[[io.buildpacks.build.env]]`

An example would look like this:

```TOML
[_]
schema-version = "0.3"

[[io.buildpacks.group]]
id = "buildpacks/ruby"
version = "latest"

[[io.buildpacks.group]]
id = "buildpacks/nodejs"
version = "latest"
exec-env = ["production", "test"]

[[io.buildpacks.group]]
id = "buildpacks/metrics-agent"
version = "latest"
exec-env = ["production"]

[[io.buildpacks.group]]
id = "buildpacks/headless-chrome"
version = "latest"
exec-env = ["test"]

[[io.buildpacks.post.group]]
id = "buildpacks/procfile"
version = "latest"

[[io.buildpacks.build.env]]
name = "RAILS_ENV"
value = "production"
exec-env = ["production"]

[[io.buildpacks.build.env]]
name = "RAILS_ENV"
value = "test"
exec-env = ["test"]

[[io.buildpacks.build.env]]
name = "PARALLEL_WORKERS"
value = "4"
exec-env = ["production"]
```

### `builder.toml` (Builder Authors)

> The the only table `exec-env` will be added to is `[[order.group]]` and `[[build.env]]`.

Reviewer 7: ```suggestion
The `exec-env` key will be added to the `[[order.group]]` and `[[build.env]]` arrays of tables.
```

### `buildpack.toml` (Buildpack Authors)

As a piece of metadata, a buildpack should list the execution environments it supports in the `[[buildpack.exec-env]]` array. This can be used by the buildpack registry in the future.

For composite buildpacks, `exec-env` will be added to is `[[buildpack.order.group]]`.

### `launch.toml` (Buildpack Authors)

Not all process types make sense for every execution environment. In order to help hint to the platform the intention how a process should be used, the `exec-env` key will be added to the `[[processes]]` table.

### `metadata.toml` (Platform Operators)

> On the platform side, the `exec-env` key will be added to `metadata.toml` in the `[[processes]]` table to mirror `launch.toml`. This will help platforms to make decisions on which processes are desired for each execution environment. `lifecycle` will list all processes and will not exclude any based on the execution environment Instead, a platform can use the available information to make a decision.

Reviewer 7: ```suggestion
On the platform side, the `exec-env` key will be added to `metadata.toml` in the `[[processes]]` table to mirror `launch.toml`. This will help platforms decide which processes are desired for each execution environment. `lifecycle` will list all processes and will not exclude any based on the execution environment Instead, a platform can use the available information to make a decision.
```

## `CNB_EXEC_ENV` Environment Variable

This value is a string and must abide by similar rules we use for IDs:

* MUST only contain numbers, letters, and the characters `.`, and `-`.

The spec will reserve the following values to help standardize execution environments:

* production
* test
* development

In addition, the `/` character is reserved in case we need to introduce namespacing in the future.

### Buildpack API

A buildpack author will be able to determine the execution environment their buildpack is expected to build for by reading the `CNB_EXEC_ENV` environment variable. If this value is not set, a Buildpack Author can assume it's set to `production`. This will be provided for both `bin/detect` and `bin/build`.

> This would let a buildpack author do different things based on the execution environment. For example, it's common for a production build to not include test dependencies or files, while a test environment would include those things. For compiled languages, a production might even remove the source code and just leave the compiled binary with optimizations. In a test environment, it may include debug symbols, not run with optimizations, not prune the source tree.

Reviewer 7: ```suggestion
This would let a buildpack author do different things based on the execution environment. For example, it's common for a production build to exclude test dependencies or files that a test environment would include. For compiled languages, a production build might even remove the source code and just leave the compiled binary with optimizations. In a test environment, a build may include debug symbols, be compiled without optimizations, and leave the source code untouched.
```

# Drawbacks
[drawbacks]: #drawbacks

While the design tries to stay simple, it still adds new additions to the spec.

* env var to Buildpack API
* env var to Platform API
* field to Project Descriptor, `builder.toml`, and `buildpack.toml`.

# Alternatives
[alternatives]: #alternatives

## Separate Execution Environment Project Descriptor files

When using `Dockerfile`, it's common to create a separate one for other execution environments like `Dockerfile.test`. This makes the file clean and easy to read, but comes at the cost of duplicating setup/code.

With Project Descriptor, this would remove the need for the `exec-env` key at the cost of replicating buildpack groups. It does bring a big benefit by easily unlocking the `builder` key in Project Descriptor to be specific to an execution environment. This would apply to any fields (not tables) that exist at the `[io.buildpacks]` level like `include`/`exclude`. There are some other options if this is desired, where a new table could be created for hoisting fields into that would allow us to add the `exec-env` field.

# Prior Art
[prior-art]: #prior-art

## [Heroku Testpack API](https://devcenter.heroku.com/articles/testpack-api)

> As part of the classic buildpack design, Heroku extended the API to include support for tests. It added a `bin/test-compile` phase, which is like the normal `bin/compile` phase but specific for setting things up for test. `bin/test` was called for executing tests. This design let a buildpack have a specific codepath for tests and also do intelligent logic for executing tests based on the source code. For example [in Ruby](https://github.com/heroku/heroku-buildpack-ruby/blob/main/bin/support/ruby_test#L62-L76), the buildpack author can run `bin/rspec` if it detects `rspec`, `bin/rails test` when using rails, or default to `rake test`.

Reviewer 7: ```suggestion
As part of the classic buildpack design, Heroku extended the API to include support for tests. It added a `bin/test-compile` phase, which is similar to the `bin/compile` phase, but specific for setting things up for test. `bin/test` was called for executing tests. This design lets buildpack authors write codepaths and logic optimized specifically for executing tests based on the source code. For example, [this Ruby buildpack](https://github.com/heroku/heroku-buildpack-ruby/blob/fb393cb46e23ab809e21daeef2a97cbd5f04a370/bin/support/ruby_test#L62-L76) will run `bin/rspec` if it detects `rspec`, `bin/rails test` when using rails, or default to `rake test`.
```
